runner/app/live/streamer/process.py

```diff
@@ -117,91 +119,108 @@ def process_loop(self):
-                    continue
+    async def _run_pipeline_loops(self):
+        pipeline = await self._initialize_pipeline()
+        await pipeline.warm_video()
+        input_task = asyncio.create_task(self._input_loop(pipeline))
+        output_task = asyncio.create_task(self._output_loop(pipeline))
+        param_task = asyncio.create_task(self._param_update_loop(pipeline))
 
-                try:
-                    if isinstance(input_frame, VideoFrame):
-                        input_frame.log_timestamps["pre_process_frame"] = time.time()
-                        output_image = pipeline.process_frame(input_frame.image)
-                        input_frame.log_timestamps["post_process_frame"] = time.time()
-                        output_frame = VideoOutput(input_frame.replace_image(output_image))
-                        self.output_queue.put(output_frame)
-                    elif isinstance(input_frame, AudioFrame):
-                        self.output_queue.put(AudioOutput([input_frame]))
-                        # TODO wire in a proper pipeline here
-                    else:
-                        report_error(f"Unsupported input frame type {type(input_frame)}")
-                except Exception as e:
-                    report_error(f"Error processing frame: {e}")
+        try:
+            await asyncio.gather(input_task, output_task, param_task)
         except Exception as e:
-            report_error(f"Error in process run method: {e}")
-        finally:
-            self._cleanup_pipeline(pipeline)
+            self._report_error(f"Error in pipeline loops: {e}")
+            await self._cleanup_pipeline(pipeline)
+            raise
+
+    async def _input_loop(self, pipeline):
+        while not self.is_done():
+            try:
+                input_frame = await asyncio.to_thread(self.input_queue.get, timeout=0.1)
+                if isinstance(input_frame, VideoFrame):
+                    input_frame.log_timestamps["pre_process_frame"] = time.time()
+                    await pipeline.put_video_frame(input_frame)
+                elif isinstance(input_frame, AudioFrame):
+                    self.output_queue.put(AudioOutput([input_frame]))
+            except queue.Empty:
+                continue
+            except Exception as e:
+                self._report_error(f"Error processing input frame: {e}")
+
+    async def _output_loop(self, pipeline):
+        while not self.is_done():
+            try:
+                output_frame = await pipeline.get_processed_video_frame()
+                output_frame.log_timestamps["post_process_frame"] = time.time()
+                await asyncio.to_thread(self.output_queue.put, output_frame)
+            except Exception as e:
+                self._report_error(f"Error processing output frame: {e}")
+
+    async def _param_update_loop(self, pipeline):
+        while not self.is_done():
+            try:
+                params = self.param_update_queue.get_nowait()
```

Comment:
We could probably have an `asyncio.to_thread` with no timeout here as well

Reply:
Addressed in https://github.com/livepeer/ai-runner/commit/d6922ab24ea8a5e22e3d565500404273be361ad2